<a href="https://colab.research.google.com/github/sarina2668/sarina2668-Sarina_INFO4670_Spring2026/blob/main/Assignment_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#1. Data Quality Check

1.1 Using Python (pandas, matplotlib, or seaborn), load and inspect the Assignment 2 dataset.

In [1]:
# Write your code here
import pandas as pd
from google.colab import files
uploaded = files.upload()



Saving Assignment 2 dataset.csv to Assignment 2 dataset.csv


Write code to explore the data distribution (e.g., region, type, year) and check whether there is any bias. Provide both the code and your interpretation.

In [2]:
# Write your code here

df = pd.read_csv("Assignment 2 dataset.csv")

df.head()



Column 1        Date  AveragePrice  Total Volume     4046       4225  \
0         0  12-27-2015          1.33      64236.62  1036.74   54454.85   
1         1  12-20-2015          1.35      54876.98   674.28   44638.81   
2         2  12-13-2015          0.93     118220.22   794.70  109149.67   
3         3   12-6-2015          1.08      78992.15  1132.00   71976.41   
4         4  11-29-2015          1.28      51039.60   941.48   43838.39   

     4770  Total Bags  Small Bags  Large Bags  XLarge Bags          type  \
0   48.16     8696.87     8603.62       93.25          0.0  conventional   
1   58.33     9505.56     9408.07       97.49          0.0  conventional   
2  130.50     8145.35     8042.21      103.14          0.0  conventional   
3   72.58     5811.16     5677.40      133.76          0.0  conventional   
4   75.78     6183.95     5986.26      197.69          0.0  conventional   

   year  region  
0  2015  Albany  
1  2015  Albany  
2  2015  Albany  
3  2015  Albany  
4  2015  Albany

In [3]:
#by region
region_counts = df["region"].value_counts()
region_counts.head(10)

#by type
type_counts = df["type"].value_counts()
type_counts

#by year
year_counts = df["year"].value_counts().sort_index()
year_counts



year
1904       1
2015    5615
2016    5616
2017    5722
2018    1300
Name: count, dtype: int64

interpretation 1: There is only one region (WestTexNewMexico) that has 2 more rows than the rest of the region, but this not a strong bias. Type does not have a strong bias as well. However, year does have a strong bias because it has 1 row from 1994, which is not close to any of the other years presented. 2018 does not reflect a close count to 2015-2018 years.

1.2 Write Python code to check for duplicate rows and missing values in the dataset. Show the number of duplicates and missing values for each column. Then, explain (in comments or markdown) how you would handle these issues (e.g., drop, impute, or replace).

In [4]:
# Write your code here
duplicate_rows = df.duplicated().sum()


missing_values = df.isna().sum()


print(duplicate_rows,',', missing_values)



2 , Column 1        0
Date            0
AveragePrice    0
Total Volume    1
4046            2
4225            1
4770            1
Total Bags      1
Small Bags      2
Large Bags      2
XLarge Bags     1
type            1
year            0
region          0
dtype: int64


Interpretation 2: Normally, in a big dataset with a significant amount of missing values, I would replace those values with 0 or the median. However, in this case, there are not that many missing values for most columns, so I would just remove them. I would also remove the duplicate values



1.3 Use Python code to print the number of rows and columns in the dataset (e.g., with df.shape). Based on the dataset size, explain (briefly) whether you think the dataset is sufficient for training a machine learning model.

In [5]:
# Write your code here
rows, columns = df.shape
print(rows, columns)



18254 14


I think this dataset is sufficient for training a machine lealrning model because it is large enough to capture patterns

#2. Data Cleaning and Preprocessing

2.1 Remove the first column or “Column 1” from the dataset. Treat the ‘year’ variable as nominal.

In [9]:
# Write your code here

#df.drop("Column 1", axis=1, inplace=True)
print(df.columns)

df["year"]= df["year"].astype("category")
df.dtypes

Index(['Date', 'AveragePrice', 'Total Volume', '4046', '4225', '4770',
       'Total Bags', 'Small Bags', 'Large Bags', 'XLarge Bags', 'type', 'year',
       'region'],
      dtype='object')


Date              object
AveragePrice     float64
Total Volume     float64
4046             float64
4225             float64
4770             float64
Total Bags       float64
Small Bags       float64
Large Bags       float64
XLarge Bags      float64
type              object
year            category
region            object
dtype: object

2.2 Check for duplicate values and remove them.

In [10]:
# Write your code here

df.drop_duplicates(inplace=True)

num_duplicates = df.duplicated()
print (num_duplicates)

0        False
1        False
2        False
3        False
4        False
         ...  
18247    False
18248    False
18250    False
18251    False
18252    False
Length: 18252, dtype: bool


2.3 Check for missing values. If a data record (row) only has a few missing values, replace the missing values with the median of the column feature in that specific “Region” variable. If most column values in a data record are missing, remove the data record.

In [11]:
# Write your code here

print(df.isnull().sum())

df = df[df.isnull().mean(axis=1) <= 0.5]

numeric_cols = df.select_dtypes(include="float64").columns

df[numeric_cols] = (
    df.groupby("region")[numeric_cols]
      .transform(lambda x: x.fillna(x.median()))
)

Date            0
AveragePrice    0
Total Volume    1
4046            2
4225            1
4770            1
Total Bags      1
Small Bags      2
Large Bags      2
XLarge Bags     1
type            1
year            0
region          0
dtype: int64


2.4 Find the correlation between the variables and describe how the correlated values among the variables impact the model accuracy.


In [12]:
# Write your code here

corr_matrix = df.select_dtypes(include="number").corr()


corr_matrix




AveragePrice  Total Volume      4046      4225      4770  \
AveragePrice      1.000000     -0.192767 -0.208325 -0.172944 -0.179458   
Total Volume     -0.192767      1.000000  0.977863  0.974181  0.872203   
4046             -0.208325      0.977863  1.000000  0.926110  0.833390   
4225             -0.172944      0.974181  0.926110  1.000000  0.887856   
4770             -0.179458      0.872203  0.833390  0.887856  1.000000   
Total Bags       -0.177103      0.963047  0.920057  0.905788  0.792315   
Small Bags       -0.174742      0.967238  0.925280  0.916032  0.802734   
Large Bags       -0.172953      0.880640  0.838646  0.810016  0.698473   
XLarge Bags      -0.117604      0.747158  0.699378  0.688810  0.679862   

              Total Bags  Small Bags  Large Bags  XLarge Bags  
AveragePrice   -0.177103   -0.174742   -0.172953    -0.117604  
Total Volume    0.963047    0.967238    0.880640     0.747158  
4046            0.920057    0.925280    0.838646     0.699378  
4225            0.905788    0.916032    0.810016     0.688810  
4770            0.792315    0.802734    0.698473     0.679862  
Total Bags      1.000000    0.994334    0.943009     0.804233  
Small Bags      0.994334    1.000000    0.902589     0.806845  
Large Bags      0.943009    0.902589    1.000000     0.710860  
XLarge Bags     0.804233    0.806845    0.710860     1.000000

They imapct the model accuracy by revealing patterns that help with analysis. For example, Volume and PLU codes have a positive corelated relationship because they increase together. Whereas, averagePrice is negatively correlated with volume features, indicating that an increase in avocados leads to lower prices.

#3. Exploratory Data Analysis (EDA)


3.1 Describe the variables
- Describe all variables in the dataset.
- For continuous variables: report **range (min, max), mean, median, and distribution**.
- For categorical variables: list unique values.

In [13]:
# Write your code here


numeric_cols = df.select_dtypes(include="number")


numeric_summary = numeric_cols.describe().T
numeric_summary["median"] = numeric_cols.median()
numeric_summary = numeric_summary[["min", "max", "mean", "median"]]
numeric_summary

cat_cols = df.select_dtypes(include='category')

for col in cat_cols:
    print(f"{col}: {df[col].unique()}")


year: [2015, 2016, 2017, 2018, 1904]
Categories (5, int64): [1904, 2015, 2016, 2017, 2018]


-AveragePrice: Prices range from $0.44 to $3.44 Distribution shows a slight skew towards higher prices


-TotalVolume: Amount of avocados ranges from 84.56 to 62,505,646.52. Distribution shows a significant skew towards the right.


-The plu numbers and Bags are skewed by high outliers. With 4770 and Xlarge bags making the least sales

-Categorical: year- Appears to be erroneous


3.2 Inspect the earliest recorded date
- Find the earliest `Date`.
- Check if there are avocado prices recorded from the earliest date up to 2010.
- Comment: does the earliest data point look reasonable? Keep or remove?

In [14]:

df["Date"] = pd.to_datetime(df["Date"])


earliest_date = df["Date"].min()
earliest_date


df[df['Date'] < '2010-01-01']

Date  AveragePrice  Total Volume     4046     4225    4770  \
18250 1904-01-21          1.87      13766.76  1191.92  2452.79  727.94   

       Total Bags  Small Bags  Large Bags  XLarge Bags     type  year  \
18250     9394.11      9351.8       42.31          0.0  organic  1904   

                 region  
18250  WestTexNewMexico

This early data point is not reasonable because all the other data points represent sales that are a century ahead. In other words, this data point is irrelevant.

3.3 Highest average price
- Find the highest value in "AveragePrice".
- Report which region it belongs to.
- Describe how you obtained the result.

In [15]:
# Write your code here

highest_price = df["AveragePrice"].max()
highest_price

highest_price_row = df[df["AveragePrice"] == highest_price]

#highest_price_row
highest_price_row[["region", "AveragePrice", "Date", "type"]]

region  AveragePrice       Date     type
14125  SanFrancisco          3.25 2016-10-30  organic

San franciso is the region with the highest average price. I used the max function to find the highest average price. Then, I located the row with the highest average.

3.4 Highest total volume
- Find the highest total volume of avocados.
- Report which region it belongs to.
- Describe how you obtained the result.

In [16]:
# Write your code here
highest_volume = df["Total Volume"].max()

highest_volume_row = df[df["Total Volume"] == highest_volume]


highest_volume_row[["region", "Total Volume", "Date", "type"]]

region  Total Volume       Date          type
9097  TotalUS   62505646.52 2018-02-04  conventional

TotalUS is the region with the highest average price. I used the max function to find the highest average price. Then, I located the row with the highest average.